README.md
# true-promise

TruePromise is a an overhaul of the javascript class Promise. [Learn how to use Promises](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise).



# Introduction

TruePromise is meant to be used with TypeScript. It may not fit your needs. Try [bluebird](https://github.com/petkaantonov/bluebird).

_TruePromise works with async/await syntax in a nodeJS environnement. ⚠️ Not tested in a commonJS environnement._

- Does not throw error if not handled.
- Allows delayed calling of then/catch.
- Correctly types the value in reject cases.
- Rework of the static methods with better types.
- New timeout() static method.

**Feel free to suggest any improvement.**



# Installation

###### NPM
```npm install --save true-promise```

###### Yarn
```yarn add true-promise```

# Importing

```ts
import TruePromise from 'true-promise';
```


# Usage

The main purpose of a Promise is to handle a script that can be accepted (resolved), or refused (rejected). For each case, it can send back a value.
TruePromise is used the same way as javascript Promise. Some returned value may change in static methods.
All methods offer documentation while hovering them in your IDE. Most of them are taken from [Mozilla's Promise documentation](https://developer.mozilla.org/fr/docs/Web/JavaScript/Reference/Global_Objects/Promise).

## New TruePromise

The difference with javascript Promise is that you can type the value that is supposed to come when rejecting the promise. This is meant to use a different rejection way than throwing an error.

```ts
import TruePromise from 'true-promise';

// Here, the promise must resolve with a string, or reject with a number.
new TruePromise<string, number>((resolve, reject) => {
    resolve("true promise"); // Works
    resolve(1); // Error
    reject("true promise"); // Error
    reject(1); // Works
});
```

## Handling a TruePromise future

Unlike javascript Promise, you don't have to call then and catch. You can also call them later.

```ts
import TruePromise from 'true-promise';

const promise = new TruePromise<string, number>((resolve, reject) => {
  reject(1); // The Promise is immediatly rejected
}); // No then() called.

setTimeout(() => {
  promise.catch((value) => console.log(value)); // Since the promise has been previously rejected, the reject function is immediatly called. And now know for sure the value is a number.
}, 1000);
```

## Static methods

Statics methods have been reworked to implement the rejected value type feature. All static methods taking promises as argument accept javascript Promise and TruePromise.

### TruePromise.all()
The Promise.all() method takes an iterable of promises as an input, and returns a single Promise that resolves to an array of the results of the input promises, in the same order. This returned promise will resolve when all of the input's promises have resolved, or if the input iterable contains no promises. It rejects immediately upon any of the input promises rejecting, and will reject with this first value as {value, index: index of the rejected promise}.

```ts
// Example immediatly resolving
TruePromise.all([])
    .then((values) => console.log('then', values)) // Immediatly log `then []`.
    .catch((values) => console.log('catch', values)); // Never called in this case.

// Example resolving after 500ms
TruePromise.all<string, number>([
    new Promise((resolve, reject) => resolve("string")),
    new TruePromise((resolve, reject) => setTimeout(() => resolve("string"), 500)),
    new TruePromise((resolve, reject) => resolve("string")),
])
    .then((values) => console.log('then', values)) // Log `then ["string", "string", "string"]` after 500ms.
    .catch((rejection) => console.log('catch', rejection)) // Never called in this case.

// Exemple rejecting after 1s
TruePromise.all<string, number>([
    new Promise((resolve, reject) => resolve("string")),
    new TruePromise((resolve, reject) => setTimeout(() => reject(2), 1000)), // Rejecting after 1s.
    new TruePromise((resolve, reject) => resolve("string")),
])
    .then((values) => console.log('then', values)) // Never called in this case.
    .catch((rejection) => console.log('catch', rejection)) // Log `catch {value: 2, index: 1}` after 1s.
```

### TruePromise.allSettled()
The Promise.allSettled() method returns a promise that resolves after all of the given promises have either resolved or rejected, with an array of objects that each describes the outcome of each promise. It is typically used when you have multiple asynchronous tasks that are not dependent on one another to complete successfully, or you'd always like to know the result of each promise. In comparison, the Promise returned by Promise.all() may be more appropriate if the tasks are dependent on each other / if you'd like to immediately reject upon any of them rejecting.

```ts
// Example immediatly resolving
TruePromise.allSettled([])
    .then((values) => console.log('then', values)) // Immediatly log `then []`.
    .catch((values) => console.log('catch', values)); // Never called in any case.

// Example resolving after 500ms
TruePromise.allSettled<string, number>([
    new Promise((resolve, reject) => resolve("string")),
    new TruePromise((resolve, reject) => setTimeout(() => resolve("string"), 500)),
    new TruePromise((resolve, reject) => resolve("string")),
])
    .then((values) => console.log('then', values)) // Log `then [{ value: 'string', status: 'resolved' },{ value: 'string', status: 'resolved' },{ value: 'string', status: 'resolved' }]` after 500ms.
    .catch((rejection) => console.log('catch', rejection)) // Never called in any case.

// Exemple resolving after 1s
TruePromise.allSettled<string, number>([
    new Promise((resolve, reject) => resolve("string")),
    new TruePromise((resolve, reject) => setTimeout(() => reject(2), 1000)), // Rejecting after 1s.
    new TruePromise((resolve, reject) => resolve("string")),
])
    .then((values) => console.log('then', values)) // Log `then [{ value: 'string', status: 'resolved' },{ value: 2, status: 'rejected' },{ value: 'string', status: 'resolved' }]` after 1s.
    .catch((rejection) => console.log('catch', rejection)) // Never called in any case.
```


### TruePromise.any()
Promise.any() takes an array of Promises and, as soon as one of the promises in the array fulfills, returns a single promise that resolves with the value from that promise. If no promises in the array fulfill (if all of the given promises are rejected), then the returned promise is rejected with all of the rejected values from the promise, in the same order.

```ts
// Example immediatly rejecting
TruePromise.any([])
    .then((values) => console.log('then', values)) // Never called in this case.
    .catch((values) => console.log('catch', values)); // Immediatly log `catch[]`.

// Example resolving after 500ms
TruePromise.any<string, number>([
    new Promise((resolve, reject) => resolve("string")),
    new TruePromise((resolve, reject) => setTimeout(() => resolve("string"), 500)),
    new TruePromise((resolve, reject) => resolve("string")),
])
    .then((values) => console.log('then', values)) // Log `then { value: 'string', index: 2 }` after 500ms.
    .catch((rejection) => console.log('catch', rejection)) // Never called in this case.

// Exemple resolving after 1s
TruePromise.any<string, number>([
    new Promise((resolve, reject) => resolve("string")),
    new TruePromise((resolve, reject) => setTimeout(() => reject(2), 1000)), // Rejecting after 1s.
    new TruePromise((resolve, reject) => resolve("string")),
])
    .then((values) => console.log('then', values)) // Log `then { value: 'string', index: 2 }` after 1s.
    .catch((rejection) => console.log('catch', rejection)) // Never called in this case.
```

### TruePromise.race()
The Promise.race() method returns a promise that fulfills or rejects as soon as one of the promises in an iterable fulfills or rejects, with the value from that promise and index of it.

```ts
// Example never resolving or rejecting
TruePromise.race([])
    .then((values) => console.log('then', values)) // Never called in this case.
    .catch((values) => console.log('catch', values)); // Never called in this case.

// Example resolving after 500ms
TruePromise.race<string, number>([
    new Promise((resolve, reject) => resolve("string")),
    new TruePromise((resolve, reject) => setTimeout(() => resolve("string"), 500)),
    new TruePromise((resolve, reject) => resolve("string")),
])
    .then((values) => console.log('then', values)) // Log `then { value: 'string', index: 2 }` after 500ms.
    .catch((rejection) => console.log('catch', rejection)) // Never called in this case.

// Exemple resolving after 1s
TruePromise.race<string, number>([
    new Promise((resolve, reject) => resolve("string")),
    new TruePromise((resolve, reject) => setTimeout(() => reject(2), 1000)), // Rejecting after 1s.
    new TruePromise((resolve, reject) => resolve("string")),
])
    .then((values) => console.log('then', values)) //  Log `then { value: 'string', index: 2 }` after 1s.
    .catch((rejection) => console.log('catch', rejection)) // Never called in this case.
```

### TruePromise.reject()
The Promise.reject() method returns a Promise object that is rejected with a given value.

```ts
// Example immedialty resolving
TruePromise.reject("string")
    .then((value) => console.log('then', value)) // Never called in this case.
    .catch((value) => console.log('catch', value)); // Immediatly log `string`.
```

### TruePromise.resolve()
The Promise.resolve() method returns a Promise object that is resolved with a given value.

```ts
// Example immedialty resolving
TruePromise.resolve("string")
    .then((value) => console.log('then', value)) // Immediatly log `string`.
    .catch((value) => console.log('catch', value)); // Never called in this case.
```

### TruePromise.timeout()
The Promise.timeout() method returns a Promise object that resolves when the given promise is resolved, and rejects when the given promise is rejected or when the given timeout is up.

```ts
// Example immedialty resolving
TruePromise.timeout<string, number>((resolve, reject) => {
    resolve("string")
}, 1000)
    .then((value) => console.log('then', value)) // Immediatly log `string`.
    .catch((value) => console.log('catch', value)); // Never called in this case.

        // Example immedialty resolving
TruePromise.timeout<string, number>((resolve, reject) => {
    setTimeout(() => resolve("string"), 2000);
}, 1000)
    .then((value) => console.log('then', value)) // Never called in this case.
    .catch((value) => console.log('catch', value)); // Log "timeout" after 1s.
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [1.0.7](https://github.com/lenacassandre/true-promises/compare/v1.0.6...v1.0.7) (2021-06-25)


### Bug Fixes

* readme ([d7af8c2](https://github.com/lenacassandre/true-promises/commit/d7af8c299e21996b88ea875ae775b005e11ab338))

### [1.0.6](https://github.com/lenacassandre/true-promises/compare/v1.0.5...v1.0.6) (2021-06-25)


### Bug Fixes

* readme ([17aa9b9](https://github.com/lenacassandre/true-promises/commit/17aa9b90fb7ff9e9595f1c7d797ca7050c398aa6))

### [1.0.5](https://github.com/lenacassandre/true-promises/compare/v1.0.4...v1.0.5) (2021-06-25)


### Bug Fixes

* fix readme ([03a0e4b](https://github.com/lenacassandre/true-promises/commit/03a0e4b3d9e3c5c95afadce139ca1489971e8599))

### [1.0.4](https://github.com/lenacassandre/true-promises/compare/v1.0.3...v1.0.4) (2021-06-25)


### Bug Fixes

* fixing exports ([2678d7f](https://github.com/lenacassandre/true-promises/commit/2678d7fe2e2bb553d2bfa8e6031446792be5a810))

### [1.0.3](https://github.com/lenacassandre/true-promises/compare/v1.0.2...v1.0.3) (2021-06-25)


### Bug Fixes

* README.md  ([6968bb7](https://github.com/lenacassandre/true-promises/commit/6968bb70b79842605a145320bdcb74c27cd11a43))

### [1.0.2](https://github.com/lenacassandre/true-promises/compare/v1.0.1...v1.0.2) (2021-06-24)


### Bug Fixes

* **package.json:** fixed repo name ([ebe07bd](https://github.com/lenacassandre/true-promises/commit/ebe07bd8621a83dee9772b86e672b2b848cf14fd))

### [1.0.1](https://github.com/lenacassandre/true-promises/compare/v0.0.4...v1.0.1) (2021-06-24)

### [0.0.4](https://github.com/lenacassandre/true-promises/compare/v0.0.3...v0.0.4) (2021-06-24)


### Bug Fixes

* **tests:** comment test ([1297f86](https://github.com/lenacassandre/true-promises/commit/1297f86f0293747506fed742a0c648c66de6494e))

### [0.0.3](https://github.com/lenacassandre/true-promises/compare/v0.0.2...v0.0.3) (2021-06-24)

### 0.0.2 (2021-06-24)


### Features

* **truepromise:** rewriting of javascript promise ([b1575c8](https://github.com/lenacassandre/true-promises/commit/b1575c8aeca4bc042cb99db6423dfb580e0721ca))


lib/classes/TruePromise.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TruePromise = void 0;
class TruePromise {
    constructor(executor) {
        this._resolve = (value) => {
            if (this._state === "pending") {
                this._state = "resolved";
                this.resolveValue = value;
                if (this.resolveCallback) {
                    this.resolveCallback(this.resolveValue);
                    this._state = "finished";
                }
            }
        };
        this._reject = (value) => {
            if (this._state === "pending") {
                this._state = "rejected";
                this.rejectValue = value;
                if (this.rejectCallback) {
                    this.rejectCallback(this.rejectValue);
                    this._state = "finished";
                }
            }
        };
        this.catch = (onReject) => {
            this.rejectCallback = onReject;
            if (this._state === "rejected") {
                this.rejectCallback(this.rejectValue);
                this._state = "finished";
            }
            return this;
        };
        this.finally = (onResolveOrReject) => {
            return this.then(onResolveOrReject).catch(onResolveOrReject);
        };
        this.then = (onResolve, onReject) => {
            this.resolveCallback = onResolve;
            if (this._state === "resolved") {
                this.resolveCallback(this.resolveValue);
                this._state = "finished";
            }
            if (onReject) {
                return this.catch(onReject);
            }
            return this;
        };
        this._state = "pending";
        executor(this._resolve, this._reject);
        return this;
    }
    get state() {
        return this._state;
    }
}
exports.TruePromise = TruePromise;
TruePromise.all = (promises) => {
    return new TruePromise((resolve, reject) => {
        if (promises.length === 0)
            return resolve([]);
        const resolveValues = [];
        let resolvedPromisesAmount = 0;
        let state = "pending";
        const addResolveValue = (value, index) => {
            resolveValues[index] = value;
            resolvedPromisesAmount++;
            if (resolvedPromisesAmount === promises.length && state === "pending") {
                state = "finished";
                resolve(resolveValues);
            }
        };
        const rejectMainPromise = (value, index) => {
            if (state === "pending") {
                state = "finished";
                reject({ value, index });
            }
        };
        promises.forEach((promise, index) => {
            promise
                .then((value) => addResolveValue(value, index))
                .catch((value) => rejectMainPromise(value, index));
        });
    });
};
TruePromise.allSettled = (promises) => {
    return new TruePromise((resolve, reject) => {
        if (promises.length === 0)
            return resolve([]);
        const values = [];
        let resolvedPromisesAmount = 0;
        let state = "pending";
        const addResolveValue = (value, promiseState, index) => {
            values[index] = { value, status: promiseState };
            resolvedPromisesAmount++;
            if (resolvedPromisesAmount === promises.length && state === "pending") {
                state = "finished";
                resolve(values);
            }
        };
        promises.forEach((promise, index) => {
            promise
                .then((value) => addResolveValue(value, "resolved", index))
                .catch((value) => addResolveValue(value, "rejected", index));
        });
    });
};
TruePromise.any = (promises) => {
    return new TruePromise((resolve, reject) => {
        if (promises.length === 0)
            return reject([]);
        const rejectedValues = [];
        let rejectedPromisesAmount = 0;
        let state = "pending";
        const addRejectedValue = (value, index) => {
            rejectedValues[index] = value;
            rejectedPromisesAmount++;
            if (rejectedPromisesAmount === promises.length && state === "pending") {
                state = "finished";
                reject(rejectedValues);
            }
        };
        const resolveMainPromise = (value, index) => {
            if (state === "pending") {
                state = "finished";
                resolve({ value, index });
            }
        };
        promises.forEach((promise, index) => {
            promise
                .then((value) => resolveMainPromise(value, index))
                .catch((value) => addRejectedValue(value, index));
        });
    });
};
TruePromise.race = (promises) => {
    return new TruePromise((resolve, reject) => {
        let state = "pending";
        const resolveMainPromise = (value, index) => {
            if (state === "pending") {
                state = "finished";
                resolve({ value, index });
            }
        };
        const rejectMainPromise = (value, index) => {
            if (state === "pending") {
                state = "finished";
                reject({ value, index });
            }
        };
        promises.forEach((promise, index) => {
            promise
                .then((value) => resolveMainPromise(value, index))
                .catch((value) => rejectMainPromise(value, index));
        });
    });
};
TruePromise.reject = (rejectValue) => {
    return new TruePromise((_resolve, reject) => reject(rejectValue));
};
TruePromise.resolve = (resolveValue) => {
    return new TruePromise((resolve, _reject) => resolve(resolveValue));
};
TruePromise.timeout = (executor, timeout) => {
    return new TruePromise((resolve, reject) => {
        new TruePromise(executor)
            .then(resolve)
            .catch(reject);
        setTimeout(() => reject("timeout"), timeout);
    });
};
//# sourceMappingURL=TruePromise.js.map

lib/classes/TruePromise.d.ts
export declare class TruePromise<ResolveValue = void, RejectValue = void> {
    private _state;
    private resolveValue?;
    private rejectValue?;
    private resolveCallback?;
    private rejectCallback?;
    constructor(executor: (resolve: (value: ResolveValue) => void, reject: (value: RejectValue) => void) => void);
    get state(): "pending" | "resolved" | "rejected" | "finished";
    _resolve: (value: ResolveValue) => void;
    _reject: (value: RejectValue) => void;
    catch: (onReject: (value: RejectValue) => void) => this;
    finally: (onResolveOrReject: (value: ResolveValue | RejectValue) => void) => this;
    then: (onResolve: (value: ResolveValue) => void, onReject?: ((value: RejectValue) => void) | undefined) => this;
    static all: <ResolveValue_1 = any, RejectValue_1 = undefined>(promises: (TruePromise<ResolveValue_1, RejectValue_1> | Promise<ResolveValue_1>)[]) => TruePromise<ResolveValue_1[], {
        value: RejectValue_1;
        index: number;
    }>;
    static allSettled: <ResolveValue_1 = void, RejectValue_1 = void>(promises: (TruePromise<ResolveValue_1, RejectValue_1> | Promise<ResolveValue_1>)[]) => TruePromise<{
        value: ResolveValue_1 | RejectValue_1;
        status: "resolved" | "rejected";
    }[], void>;
    static any: <ResolveValue_1 = undefined, RejectValue_1 = any>(promises: (TruePromise<ResolveValue_1, RejectValue_1> | Promise<ResolveValue_1>)[]) => TruePromise<{
        value: ResolveValue_1;
        index: number;
    }, RejectValue_1[]>;
    static race: <ResolveValue_1 = void, RejectValue_1 = void>(promises: (TruePromise<ResolveValue_1, RejectValue_1> | Promise<ResolveValue_1>)[]) => TruePromise<{
        value: ResolveValue_1;
        index: number;
    }, {
        value: RejectValue_1;
        index: number;
    }>;
    static reject: <RejectValue_1 = void>(rejectValue: RejectValue_1) => TruePromise<void, RejectValue_1>;
    static resolve: <ResolveValue_1 = void>(resolveValue: ResolveValue_1) => TruePromise<ResolveValue_1, void>;
    static timeout: <ResolveValue_1 = void, RejectValue_1 = void>(executor: (resolve: (value: ResolveValue_1) => void, reject: (value: RejectValue_1) => void) => void, timeout: number) => TruePromise<ResolveValue_1, RejectValue_1 | "timeout">;
}
//# sourceMappingURL=TruePromise.d.ts.map

src/index.ts
import { TruePromise, TruePromise as Promise } from "./classes/TruePromise";

export { TruePromise, Promise };
export default TruePromise;

/* TESTS
console.clear();
console.log("")
console.log("✨ true-promises tests ✨")
console.log("")


///////////////////////////////////////////////////
// SIMPLE TEST

function simpleTest(v: number) {
    console.log("");
    console.log("========================================");
    console.log("simple test", v);

    const test = new TruePromise<string, string>((resolve, reject) => {
        console.log("simple test, inside promise", v)

        if(v === 1 ) {
            resolve('resolved')
        }
        else {
            reject('rejected')
        }
    });

    setTimeout(() => {
        test
            .then((value) => {
                console.log("then value :", value)
            })
            .catch((value) => {
                console.log("catch value :", value)
            })
    }, 500);
}

setTimeout(() => {
    simpleTest(1);
}, 0);


setTimeout(() => {
    simpleTest(2);
}, 1000);

///////////////////////////////////////////////////
// ASYNC/AWAIT TEST

setTimeout(() => {
    console.log("");
    console.log("========================================");
    console.log("Async/await tests");

    async function asyncAwaitTest(v: number) {
        console.log("");
        console.log("========================================");
        console.log("testFunction", v)

        try {
            const value = await new TruePromise<string, string>((resolve, reject) => {
                console.log("inside promise")

                if(v === 1) {
                    setTimeout(() => resolve("resolved"), 500);
                }
                else {
                    setTimeout(() => reject("rejected"), 500);
                }
            })

            console.log("value after await :", value)
        }
        catch(e) {
            console.log("catch value :", e)
        }
    }

    setTimeout(() => {
        asyncAwaitTest(1);
    }, 0);


    setTimeout(() => {
        asyncAwaitTest(2);
    }, 1000);
}, 2000);

//////////////////////////////////////////////////////////////////:
// ALL() test

setTimeout(() => {
    console.log("")
    console.log("========================================");
    console.log("All()")

    // Example immediatly resolving
    TruePromise.all([])
        .then((values) => console.log('then', values)) // Immediatly log `then []`.
        .catch((values) => console.log('catch', values)); // Never called in this case.

    // Example resolving after 500ms
    TruePromise.all<string, number>([
        new Promise((resolve, reject) => resolve("string")),
        new TruePromise((resolve, reject) => setTimeout(() => resolve("string"), 500)),
        new TruePromise((resolve, reject) => resolve("string")),
    ])
        .then((values) => console.log('then', values)) // Log `then ["string", "string", "string"]` after 500ms.
        .catch((rejection) => console.log('catch', rejection)) // Never called in this case.

    // Exemple rejecting after 1s
    TruePromise.all<string, number>([
        new Promise((resolve, reject) => resolve("string")),
        new TruePromise((resolve, reject) => setTimeout(() => reject(2), 1000)), // Rejecting after 1s.
        new TruePromise((resolve, reject) => resolve("string")),
    ])
        .then((values) => console.log('then', values)) // Never called in this case.
        .catch((rejection) => console.log('catch', rejection)) // Log `catch {value: 2, index: 1}` after 1s.
}, 4000);

//////////////////////////////////////////////////////////////////:
// ALLSETTLED() test

setTimeout(() => {
    console.log("")
    console.log("========================================");
    console.log("AllSettled()")

    // Example immediatly resolving
    TruePromise.allSettled([])
        .then((values) => console.log('then', values)) // Immediatly log `then []`.
        .catch((values) => console.log('catch', values)); // Never called in any case.

    // Example resolving after 500ms
    TruePromise.allSettled<string, number>([
        new Promise((resolve, reject) => resolve("string")),
        new TruePromise((resolve, reject) => setTimeout(() => resolve("string"), 500)),
        new TruePromise((resolve, reject) => resolve("string")),
    ])
        .then((values) => console.log('then', values)) // Log `then [{ value: 'string', status: 'resolved' },{ value: 'string', status: 'resolved' },{ value: 'string', status: 'resolved' }]` after 500ms.
        .catch((rejection) => console.log('catch', rejection)) // Never called in any case.

    // Exemple resolving after 1s
    TruePromise.allSettled<string, number>([
        new Promise((resolve, reject) => resolve("string")),
        new TruePromise((resolve, reject) => setTimeout(() => reject(2), 1000)), // Rejecting after 1s.
        new TruePromise((resolve, reject) => resolve("string")),
    ])
        .then((values) => console.log('then', values)) // Log `then [{ value: 'string', status: 'resolved' },{ value: 2, status: 'rejected' },{ value: 'string', status: 'resolved' }]` after 1s.
        .catch((rejection) => console.log('catch', rejection)) // Never called in any case.
}, 6000);

//////////////////////////////////////////////////////////////////:
// ANY() test

setTimeout(() => {
    console.log("")
    console.log("========================================");
    console.log("Any()")

    // Example immediatly rejecting
    TruePromise.any([])
        .then((values) => console.log('then', values)) // Never called in this case.
        .catch((values) => console.log('catch', values)); // Immediatly log `catch[]`.

    // Example resolving after 500ms
    TruePromise.any<string, number>([
        new Promise((resolve, reject) => resolve("string")),
        new TruePromise((resolve, reject) => setTimeout(() => resolve("string"), 500)),
        new TruePromise((resolve, reject) => resolve("string")),
    ])
        .then((values) => console.log('then', values)) // Log `then { value: 'string', index: 2 }` after 500ms.
        .catch((rejection) => console.log('catch', rejection)) // Never called in this case.

    // Exemple resolving after 1s
    TruePromise.any<string, number>([
        new Promise((resolve, reject) => resolve("string")),
        new TruePromise((resolve, reject) => setTimeout(() => reject(2), 1000)), // Rejecting after 1s.
        new TruePromise((resolve, reject) => resolve("string")),
    ])
        .then((values) => console.log('then', values)) // Log `then { value: 'string', index: 2 }` after 1s.
        .catch((rejection) => console.log('catch', rejection)) // Never called in this case.
}, 8000);

//////////////////////////////////////////////////////////////////:
// RACE() test

setTimeout(() => {
    console.log("")
    console.log("========================================");
    console.log("Race()")

    // Example never resolving or rejecting
    TruePromise.race([])
        .then((values) => console.log('then', values)) // Never called in this case.
        .catch((values) => console.log('catch', values)); // Never called in this case.

    // Example resolving after 500ms
    TruePromise.race<string, number>([
        new Promise((resolve, reject) => resolve("string")),
        new TruePromise((resolve, reject) => setTimeout(() => resolve("string"), 500)),
        new TruePromise((resolve, reject) => resolve("string")),
    ])
        .then((values) => console.log('then', values)) // Log `then { value: 'string', index: 2 }` after 500ms.
        .catch((rejection) => console.log('catch', rejection)) // Never called in this case.

    // Exemple resolving after 1s
    TruePromise.race<string, number>([
        new Promise((resolve, reject) => resolve("string")),
        new TruePromise((resolve, reject) => setTimeout(() => reject(2), 1000)), // Rejecting after 1s.
        new TruePromise((resolve, reject) => resolve("string")),
    ])
        .then((values) => console.log('then', values)) //  Log `then { value: 'string', index: 2 }` after 1s.
        .catch((rejection) => console.log('catch', rejection)) // Never called in this case.
}, 10000);

/////////////////////////////////////////////////////////////////////////
// REJECT TEST

setTimeout(() => {
    console.log("")
    console.log("========================================");
    console.log("Reject()")

    // Example immedialty resolving
    TruePromise.reject("string")
        .then((value) => console.log('then', value)) // Never called in this case.
        .catch((value) => console.log('catch', value)); // Immediatly log `string`.

}, 12000);

/////////////////////////////////////////////////////////////////////////
// RESOLVE TEST

setTimeout(() => {
    console.log("")
    console.log("========================================");
    console.log("Resolve()")

    // Example immedialty resolving
    TruePromise.resolve("string")
        .then((value) => console.log('then', value)) // Immediatly log `string`.
        .catch((value) => console.log('catch', value)); // Never called in this case.

}, 12500);


/////////////////////////////////////////////////////////////////////////
// TIMEOUT TEST

setTimeout(() => {
    console.log("")
    console.log("========================================");
    console.log("Timeout()")

    // Example immedialty resolving
    TruePromise.timeout<string, number>((resolve, reject) => {
        resolve("string")
    }, 1000)
        .then((value) => console.log('then', value)) // Immediatly log `string`.
        .catch((value) => console.log('catch', value)); // Never called in this case.

            // Example immedialty resolving
    TruePromise.timeout<string, number>((resolve, reject) => {
        setTimeout(() => resolve("string"), 2000);
    }, 1000)
        .then((value) => console.log('then', value)) // Never called in this case.
        .catch((value) => console.log('catch', value)); // Log "timeout" after 1s.

}, 13500);
*/